overlord/registrystate/registrymgr.go

```diff
@@ -37,3 +63,118 @@ func setupRegistryHook(st *state.State, snapName, hookName string, ignoreError b
 	task := hookstate.HookTask(st, summary, hookSup, nil)
 	return task
 }
+
+type changeViewHandler struct {
+	ctx *hookstate.Context
+}
+
+func (h *changeViewHandler) Before() error                 { return nil }
+func (h *changeViewHandler) Error(err error) (bool, error) { return false, nil }
+
+func (h *changeViewHandler) Done() error {
+	h.ctx.Lock()
+	defer h.ctx.Unlock()
+
+	t, _ := h.ctx.Task()
+	tx, _, err := GetStoredTransaction(t)
+	if err != nil {
+		return fmt.Errorf("cannot get transaction in change-registry handler: %v", err)
+	}
+
+	if tx.aborted() {
+		return fmt.Errorf("cannot change registry %s/%s: %s rejected changes: %s", tx.RegistryAccount, tx.RegistryName, tx.abortingSnap, tx.abortReason)
+	}
+
+	return nil
+}
+
+type saveViewHandler struct {
+	ctx *hookstate.Context
+}
+
+func (h *saveViewHandler) Before() error { return nil }
+
+func (h *saveViewHandler) Error(origErr error) (ignoreErr bool, err error) {
+	h.ctx.Lock()
+	defer h.ctx.Unlock()
+
+	t, _ := h.ctx.Task()
+	st := h.ctx.State()
+
+	var commitTaskID string
+	if err := t.Get("commit-task", &commitTaskID); err != nil {
+		return false, err
+	}
+
+	// create roll back tasks for the previously done save-registry hooks (starting
+	// with the hook that failed, so it tries to overwrite with a pristine databag
+	// just like any previous save-view hooks)
+	last := t
+	for curTask := t; curTask.Kind() == "run-hook"; curTask = curTask.WaitTasks()[0] {
+		var hooksup hookstate.HookSetup
+		if err := curTask.Get("hook-setup", &hooksup); err != nil {
+			return false, err
+		}
+
+		if !strings.HasPrefix(hooksup.Hook, "save-view-") {
+			break
+		}
+
+		// if we fail to rollback, there's nothing we can do
+		ignoreError := true
+		rollbackTask := setupRegistryHook(st, hooksup.Snap, hooksup.Hook, ignoreError)
+		rollbackTask.Set("commit-task", commitTaskID)
+		rollbackTask.WaitFor(last)
+		curTask.Change().AddTask(rollbackTask)
+		last = rollbackTask
+
+		// should never happen but let's be careful for now
+		if len(curTask.WaitTasks()) != 1 {
+			return false, fmt.Errorf("internal error: cannot rollback failed save-view: expected one prerequisite task")
+		}
+	}
+
+	// prevent the next registry tasks from running before the rollbacks. Once the
+	// last rollback task errors, these will be put on Hold by the usual mechanism
+	for _, halt := range t.HaltTasks() {
+		halt.WaitFor(last)
```

Comment:
all the run hooks will also end up with HoldStatus ?

Reply:
Once the last rollback task surfaces the error, the tasks after it will be held I believe

Reply:
will we have  a test that actually simulate running the hooks (I think MockRunHook can be used for that?) ?  that test should check the status at the end of all the tasks in the change